<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Florian Privé" />

<meta name="date" content="2017-10-21" />

<title>Analysis results in a consistent data structure</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->





<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Genomic Data Analysis in R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Pipelines</li>
    <li>
      <a href="all-in-R.html">Doing all your analysis in R</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Data Wrangling and Analysis</li>
    <li>
      <a href="eQTLanalysis.html">(Not finished) Simplified eQTL analysis in R</a>
    </li>
    <li>
      <a href="tidyverse-list-columns.html">Analysis results in a consistent data structure</a>
    </li>
    <li>
      <a href="memory-mapping.html">Genomic data analysis with large matrices in R</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Data Vizualization</li>
    <li>
      <a href="PCA.html">Unsupervised clustering</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Reporting</li>
    <li class="dropdown-header">(Coming soon) Yisong's tutorial (rmarkdown)</li>
    <li>
      <a href="xlsx_tutorial.html">Integration of Genomic Analytic Results as a Multi-tab Excel Sheet with Package 'xlsx'</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://www.hackseq.com/">hackseq</a>
</li>
<li>
  <a href="presentation/hackseq.html">Presentation</a>
</li>
<li>
  <a href="paper.html">Paper</a>
</li>
<li>
  <a href="https://github.com/hackseq/2017_project_5">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Analysis results in a consistent data structure</h1>
<h4 class="author"><em>Florian Privé</em></h4>
<h4 class="date"><em>October 21, 2017</em></h4>

</div>


<p>How do you store results of an analysis?</p>
<p>When you have different parameters that are varying, when you compare many methods and when you want to keep all the results of an analysis, your code can become quite complex.</p>
<p>In this tutorial, we make a comparison of machine learning methods for predicting disease based on small SNP data. We show how to use the tidyverse set of packages to make the analysis easier by using consistent data structures and functional programming. We use tibbles with <a href="http://r4ds.had.co.nz/many-models.html">list-columns</a>.</p>
<div id="a-simulation-made-easy" class="section level2">
<h2>A simulation made easy</h2>
<div id="data" class="section level3">
<h3>Data</h3>
<p>Let’s use some data from <a href="https://github.com/gabraham/flashpca/tree/master/HapMap3">Gad Abraham’s GitHub repository</a> which have been filtered and quality-controled.</p>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>## -- Attaching packages ----------------------------------------------------------- tidyverse 1.2.1 --</code></pre>
<pre><code>## v ggplot2 3.1.0     v purrr   0.2.5
## v tibble  2.0.1     v dplyr   0.7.8
## v tidyr   0.8.1     v stringr 1.3.1
## v readr   1.3.1     v forcats 0.3.0</code></pre>
<pre><code>## Warning: package &#39;tibble&#39; was built under R version 3.5.2</code></pre>
<pre><code>## Warning: package &#39;readr&#39; was built under R version 3.5.2</code></pre>
<pre><code>## -- Conflicts -------------------------------------------------------------- tidyverse_conflicts() --
## x dplyr::filter() masks stats::filter()
## x dplyr::lag()    masks stats::lag()</code></pre>
<pre class="r"><code>tmpfile &lt;- tempfile()
base &lt;- paste0(
  &quot;https://github.com/gabraham/flashpca/raw/master/HapMap3/&quot;,
  &quot;1kg.ref.phase1_release_v3.20101123_thinned_autosomal_overlap&quot;)
exts &lt;- c(&quot;.bed&quot;, &quot;.bim&quot;, &quot;.fam&quot;)
walk2(paste0(base, exts), paste0(tmpfile, exts), ~download.file(.x, .y))</code></pre>
<pre class="r"><code>library(bigsnpr)</code></pre>
<pre><code>## Loading required package: bigstatsr</code></pre>
<pre class="r"><code>rdsfile &lt;- snp_readBed(paste0(tmpfile, &quot;.bed&quot;))</code></pre>
<pre><code>## Warning in snp_readBed(paste0(tmpfile, &quot;.bed&quot;)): EOF of bedfile has not been reached.</code></pre>
<pre class="r"><code>bigsnp &lt;- snp_attach(rdsfile)
G &lt;- bigsnp$genotypes
G[which(is.na(G[]))] &lt;- 0  # quick and dirty imputation
obj.svd &lt;- snp_autoSVD(G, bigsnp$map$chromosome, bigsnp$map$physical.pos, ncores = 2)</code></pre>
<pre><code>## Phase of clumping (on MAF) at r^2 &gt; 0.2.. keep 14024 SNPs.
## 
## Iteration 1:
## Computing SVD..
## 0 outlier detected..
## 
## Converged!</code></pre>
<pre class="r"><code># Matrix of genotypes and first 10 PCs
X &lt;- cbind(G[], obj.svd$u)</code></pre>
</div>
<div id="methods-functions" class="section level3">
<h3>Methods’ functions</h3>
<p>Each method’s function returns a tibble (data frame) with 4 columns:</p>
<ol style="list-style-type: decimal">
<li>the name of the method,</li>
<li>the predictive scores and true phenotypes for the test set, as a <a href="http://r4ds.had.co.nz/many-models.html">list-column</a>,</li>
<li>the timing of the main computations (in seconds),</li>
<li>the number of SNPs used for the prediction.</li>
</ol>
<pre class="r"><code>logit &lt;- function(X, y, ind.train, ind.test) {
  
  timing &lt;- system.time({
    train &lt;- glmnet::cv.glmnet(X[ind.train, ], y[ind.train], type.measure = &quot;auc&quot;, 
                               family = &quot;binomial&quot;, alpha = 0.5)
    preds &lt;- predict(train, X[ind.test, ])
  })[3]
  
  tibble(
    method = &quot;logit&quot;,
    eval = list(cbind(preds, y[ind.test])),
    timing = timing,
    nb.preds = sum(glmnet::coef.cv.glmnet(train) != 0)
  )
}</code></pre>
<pre class="r"><code>svm &lt;- function(X, y, ind.train, ind.test) {
  
  timing &lt;- system.time({
    train &lt;- LiblineaR::LiblineaR(X[ind.train, ], y[ind.train], type = 5)
    preds &lt;- predict(train, X[ind.test, ], decisionValues = TRUE)
  })[3]
  
  tibble(
    method = &quot;svm&quot;,
    eval = list(cbind(1 - preds$decisionValues[, 1], y[ind.test])),
    timing = timing,
    nb.preds = sum(train$W != 0)
  )
}</code></pre>
<pre class="r"><code>xgb &lt;- function(X, y, ind.train, ind.test) {
  
  timing &lt;- system.time({
    train &lt;- xgboost::xgboost(data = X[ind.train, ], label = y[ind.train],
                              nrounds = 10, max_depth = 3, eta = 1,
                              objective = &quot;binary:logistic&quot;, 
                              verbose = FALSE, save_period = NULL)
    preds &lt;- predict(train, X[ind.test, ])
  })[3]
  
  tibble(
    method = &quot;xgb&quot;,
    eval = list(cbind(preds, y[ind.test])),
    timing = timing,
    nb.preds = nrow(xgboost::xgb.importance(model = train))
  )
}</code></pre>
</div>
<div id="simulation-of-phenotypes" class="section level3">
<h3>Simulation of phenotypes</h3>
<pre class="r"><code>get_pheno &lt;- function(
  G,                                        ## matrix of genotypes
  h2,                                       ## heritability 
  M,                                        ## nbs of causal variants
  effects.dist = c(&quot;gaussian&quot;, &quot;laplace&quot;),  ## distribution of effects 
  K = 0.3                                   ## prevalence
) {
  
  effects.dist  &lt;- match.arg(effects.dist)
  
  set &lt;- sample(ncol(G), size = M)
  effects &lt;- `if`(effects.dist == &quot;gaussian&quot;, 
                  rnorm(M, sd = sqrt(h2 / M)),
                  rmutil::rlaplace(M, s = sqrt(h2 / (2*M))))
  
  y.simu &lt;- scale(G[, set]) %*% effects
  y.simu &lt;- y.simu / sd(y.simu) * sqrt(h2)
  stopifnot(all.equal(drop(var(y.simu)), h2))
  y.simu &lt;- y.simu + rnorm(nrow(G), sd = sqrt(1 - h2))
  # Liability Threshold Model (LTM)
  pheno &lt;- as.numeric(y.simu &gt; qnorm(1 - K))
}</code></pre>
</div>
<div id="simulations" class="section level3">
<h3>Simulations</h3>
<p>First, we use <code>expand.grid()</code> to create a data frame with all possible combination of parameters we compare.</p>
<pre class="r"><code>results &lt;- expand.grid(
  M        = c(30, 3000), 
  dist     = c(&quot;gaussian&quot;, &quot;laplace&quot;), 
  num.simu = 1:10,
  stringsAsFactors = FALSE
) %&gt;% 
  as_tibble() %&gt;%
  print()</code></pre>
<pre><code>## # A tibble: 40 x 3
##        M dist     num.simu
##    &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt;
##  1    30 gaussian        1
##  2  3000 gaussian        1
##  3    30 laplace         1
##  4  3000 laplace         1
##  5    30 gaussian        2
##  6  3000 gaussian        2
##  7    30 laplace         2
##  8  3000 laplace         2
##  9    30 gaussian        3
## 10  3000 gaussian        3
## # ... with 30 more rows</code></pre>
<p>Then, we use package {purrr} to add a new column to the data frame with all the results. The main condition for a data frame is that each column has the same number of elements. So, by using lists, we can basically store any types of data in a data frame. For example, in column <code>&quot;eval&quot;</code> we store matrices, this is called a list-column. In the newly-created column <code>&quot;res&quot;</code>, we store data frames! We use tibbles instead of standard data frames because they print better.</p>
<pre class="r"><code>results[[&quot;res&quot;]] &lt;- pmap(results, function(M, dist, num.simu) {
  
  cat(&quot;.&quot;)
  
  y &lt;- get_pheno(X, 0.8, M, dist)
  
  n &lt;- nrow(X)
  ind.train &lt;- sort(sample(n, size = 500))
  ind.test &lt;- setdiff(1:n, ind.train)
  
  bind_rows(
    logit(X, y, ind.train, ind.test),
    svm(X, y, ind.train, ind.test),
    xgb(X, y, ind.train, ind.test)
  ) 
})</code></pre>
<pre><code>## ........................................</code></pre>
<pre class="r"><code>results</code></pre>
<pre><code>## # A tibble: 40 x 4
##        M dist     num.simu res             
##    &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt; &lt;list&gt;          
##  1    30 gaussian        1 &lt;tibble [3 x 4]&gt;
##  2  3000 gaussian        1 &lt;tibble [3 x 4]&gt;
##  3    30 laplace         1 &lt;tibble [3 x 4]&gt;
##  4  3000 laplace         1 &lt;tibble [3 x 4]&gt;
##  5    30 gaussian        2 &lt;tibble [3 x 4]&gt;
##  6  3000 gaussian        2 &lt;tibble [3 x 4]&gt;
##  7    30 laplace         2 &lt;tibble [3 x 4]&gt;
##  8  3000 laplace         2 &lt;tibble [3 x 4]&gt;
##  9    30 gaussian        3 &lt;tibble [3 x 4]&gt;
## 10  3000 gaussian        3 &lt;tibble [3 x 4]&gt;
## # ... with 30 more rows</code></pre>
</div>
<div id="results" class="section level3">
<h3>Results</h3>
<p>First, we use function <code>unnest()</code> of package {tidyr} so that we have a tidier data frame. To learn more about <em>tidy data</em>, please refer to <a href="http://tidyr.tidyverse.org/articles/tidy-data.html">this vignette</a> and <a href="http://r4ds.had.co.nz/tidy-data.html">this book chapter</a>.</p>
<pre class="r"><code>(results &lt;- unnest(results, res))</code></pre>
<pre><code>## # A tibble: 120 x 7
##        M dist     num.simu method eval            timing nb.preds
##    &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;  &lt;list&gt;           &lt;dbl&gt;    &lt;int&gt;
##  1    30 gaussian        1 logit  &lt;dbl [592 x 2]&gt;  11.3       105
##  2    30 gaussian        1 svm    &lt;dbl [592 x 2]&gt;   1.30      834
##  3    30 gaussian        1 xgb    &lt;dbl [592 x 2]&gt;   2.01       67
##  4  3000 gaussian        1 logit  &lt;dbl [592 x 2]&gt;  12.1         1
##  5  3000 gaussian        1 svm    &lt;dbl [592 x 2]&gt;   1.31      981
##  6  3000 gaussian        1 xgb    &lt;dbl [592 x 2]&gt;   1.82       67
##  7    30 laplace         1 logit  &lt;dbl [592 x 2]&gt;  10.6         9
##  8    30 laplace         1 svm    &lt;dbl [592 x 2]&gt;   1.30      862
##  9    30 laplace         1 xgb    &lt;dbl [592 x 2]&gt;   1.81       64
## 10  3000 laplace         1 logit  &lt;dbl [592 x 2]&gt;  11.7        61
## # ... with 110 more rows</code></pre>
<p>To compute the AUC (a measure of prediction accuracy), we use {purrr} again and specify the type of the output we want (<code>_dbl</code> for doubles). Here, we could use function <code>mutate()</code> of package {dplyr} instead of standard accessor <code>[[</code>.</p>
<pre class="r"><code>results[[&quot;AUC&quot;]] &lt;- map_dbl(results[[&quot;eval&quot;]], ~ bigstatsr::AUC(.x[, 1], .x[, 2]))
results</code></pre>
<pre><code>## # A tibble: 120 x 8
##        M dist     num.simu method eval            timing nb.preds   AUC
##    &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;  &lt;list&gt;           &lt;dbl&gt;    &lt;int&gt; &lt;dbl&gt;
##  1    30 gaussian        1 logit  &lt;dbl [592 x 2]&gt;  11.3       105 0.807
##  2    30 gaussian        1 svm    &lt;dbl [592 x 2]&gt;   1.30      834 0.671
##  3    30 gaussian        1 xgb    &lt;dbl [592 x 2]&gt;   2.01       67 0.568
##  4  3000 gaussian        1 logit  &lt;dbl [592 x 2]&gt;  12.1         1 0.5  
##  5  3000 gaussian        1 svm    &lt;dbl [592 x 2]&gt;   1.31      981 0.449
##  6  3000 gaussian        1 xgb    &lt;dbl [592 x 2]&gt;   1.82       67 0.467
##  7    30 laplace         1 logit  &lt;dbl [592 x 2]&gt;  10.6         9 0.833
##  8    30 laplace         1 svm    &lt;dbl [592 x 2]&gt;   1.30      862 0.731
##  9    30 laplace         1 xgb    &lt;dbl [592 x 2]&gt;   1.81       64 0.662
## 10  3000 laplace         1 logit  &lt;dbl [592 x 2]&gt;  11.7        61 0.553
## # ... with 110 more rows</code></pre>
<p>Now, let’s get some summaries of the results with package dplyr.</p>
<pre class="r"><code>results %&gt;%
  group_by(M, dist, method) %&gt;%
  summarise_at(c(&quot;timing&quot;, &quot;nb.preds&quot;, &quot;AUC&quot;), mean)</code></pre>
<pre><code>## # A tibble: 12 x 6
## # Groups:   M, dist [?]
##        M dist     method timing nb.preds   AUC
##    &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;   &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;
##  1    30 gaussian logit   10.4      33.8 0.800
##  2    30 gaussian svm      1.22    860   0.532
##  3    30 gaussian xgb      1.84     65.4 0.647
##  4    30 laplace  logit   10.3       9.2 0.832
##  5    30 laplace  svm      1.21    889.  0.621
##  6    30 laplace  xgb      1.81     63.6 0.674
##  7  3000 gaussian logit   11.2      71   0.546
##  8  3000 gaussian svm      1.20    968.  0.491
##  9  3000 gaussian xgb      1.81     67.3 0.511
## 10  3000 laplace  logit   11.2      80.7 0.548
## 11  3000 laplace  svm      1.23    940.  0.518
## 12  3000 laplace  xgb      1.81     67.8 0.523</code></pre>
</div>
<div id="visualise-results" class="section level3">
<h3>Visualise results</h3>
<p>Finally, let’s use package {ggplot2}, also part of the tidyverse, to visualize results.</p>
<pre class="r"><code># Wrapper to use my own theme
myggplot &lt;- function(...) ggplot(...) + bigstatsr::theme_bigstatsr()</code></pre>
<pre class="r"><code>myggplot(results) + 
  geom_boxplot(aes(method, timing)) +
  scale_y_continuous(breaks = seq(0, 20, 2)) +
  facet_grid(M ~ dist) +
  ggtitle(&quot;Timings&quot;)</code></pre>
<p><img src="tidyverse-list-columns_files/figure-html/unnamed-chunk-14-1.png" width="75%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>myggplot(results) + 
  geom_hline(yintercept = 0.5, color = &quot;red&quot;, linetype = 2) + 
  geom_boxplot(aes(method, AUC)) +
  scale_y_continuous(breaks = 0:10 / 10) +
  facet_grid(M ~ dist) +
  ggtitle(&quot;Predictive performance&quot;)</code></pre>
<p><img src="tidyverse-list-columns_files/figure-html/unnamed-chunk-15-1.png" width="75%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>The main point of using tibbles to store results is that you have the inputs, raw outputs and transformed outputs in the same data structure. Moreover, this data structure contains information that you can easily manipulate and visualize. Finally, note that we did not use a single loop in the whole method comparison.</p>
<p>The tidyverse set of packages is very convenient and consistent. You can learn more by reading <a href="http://r4ds.had.co.nz/">book R for Data Science</a>, freely available online.</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
